### PR TITLE
DAOS-6798 pool: fix deadlock on pool destroy

### PR DESCRIPTION
**Pull request by NiuYawei**
There is a deadlock on pool destroy:

1. ds_pool_stop() is called xstream 0 to put the last reference of
   ds_pool;
2. On last reference of ds_pool put, it holds 'pool_cache_lock' and
   does a collective call to clear ds_pool_child on each VOS xstream;
   (see ds_pool_put() -> pool_free_ref())
3. VOS xstreams needs to wait for aggregation ULTs exited before
   clearing the ds_pool_child; (see pool_child_delete_one() ->
   ds_cont_child_stop_all() -> cont_child_stop())
4. Aggregation ULT may running in EC aggregation code to fetch cont
   IV, and that requires calling ds_pool_lookup(), ds_pool_lookup()
   needs to acquire 'pool_cache_lock';

This patch removed 'pool_cache_lock' and ensured ds_pool cache being
accessed by xstream 0 only. If VOS xstreams need to access ds_pool,
it either uses the 'ds_pool_child->spc_pool' directly or schedules a
ULT on xstream 0.

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### CODE REVIEW

File: src/pool/srv_target.c [hunk 1]
```diff
@@ -257,13 +257,18 @@ pool_child_delete_one(void *uuid)
 	ds_pool_child_put(child); /* -1 for the list */
 
 	ds_pool_child_put(child); /* -1 for lookup */
+
+	/*
+	 * FIXME: Need to wait for last reference of ds_pool_child dropped,
+	 * since the ds_pool_child references ds_pool by 'spc_pool' without
+	 * holding ds_pool refcount.
+	 */
```
Comment (NiuYawei) [inline src/pool/srv_target.c:265]: Let's fix this in later follow-on patches.

The major use cases of 'spc_pool' is from DTX & EC aggregation (accessing ds_pool->sp_map), now we can guarantee all aggregation ULTs have already exited here. @Nasf-Fan what about DTX?

File: src/pool/srv_target.c [hunk 2]
```diff
@@ -592,17 +598,18 @@ ds_pool_start(uuid_t uuid)
...
-			/* Restart it and hold the reference. */
-			pool->sp_stopping = 0;
-		else
-			/* Already started; drop our reference. */
-			daos_lru_ref_release(pool_cache, &pool->sp_entry);
-		goto out_lock;
+		if (pool->sp_stopping) {
+			D_ERROR(DF_UUID": stopping isn't done yet\n",
+				DP_UUID(uuid));
+			rc = -DER_BUSY;
```
Comment (wangdi1) [inline src/pool/srv_target.c:604]: This seems changing the behavior.   Do we understand why it just reset sp_stopping in the original implementation.

**Review by wangdi1 — APPROVED** [review]

File: src/common/lru.c [hunk 3]
```diff
@@ -251,7 +251,10 @@ daos_lru_ref_hold(struct daos_lru_cache *lcache, void *key,
 
 	rc = d_hash_rec_insert(&lcache->dlc_htable, key, key_size,
 			       &llink->ll_link, true);
-	D_ASSERT(rc == 0);
+	if (rc == -DER_EXIST) {
+		lcache->dlc_ops->lop_free_ref(llink);
+		return rc;
+	}
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/common/lru.c:257]: indeed, will cleanup it if the patch needs be refreshed.

File: src/gurt/hash.c [hunk 4]
```diff
@@ -475,8 +475,10 @@ d_hash_rec_insert(struct d_hash_table *htable, const void *key,
 
 	if (exclusive) {
 		tmp = ch_rec_find(htable, bucket, key, ksize, D_HASH_LRU_NONE);
-		if (tmp)
+		if (tmp) {
+			D_ERROR("Dup key detected\n");
```
Comment (liw) [inline src/gurt/hash.c:479]: A D_DEBUG would be better. For many cases, -DER_EXIST and -DER_NONEXIST are not real errors, depending on callers' protocols.

File: src/pool/srv_target.c (hunk 2 above)
Comment (liw) [reply src/pool/srv_target.c:604]: Yes, this should be okay.

File: src/pool/srv_target.c [hunk 5]
```diff
@@ -1003,11 +995,11 @@ ds_pool_tgt_connect(struct ds_pool *pool, struct pool_iv_conn *pic)
 	if (hdl == NULL)
 		D_GOTO(out, rc = -DER_NOMEM);
 
-	ds_pool_get(pool);
 	uuid_copy(hdl->sph_uuid, pic->pic_hdl);
 	hdl->sph_flags = pic->pic_flags;
 	hdl->sph_sec_capas = pic->pic_capas;
-	hdl->sph_pool = pool;
+	hdl->sph_pool = ds_pool_lookup(pool->sp_uuid);
+	D_ASSERT(hdl->sph_pool != NULL);
```
Comment (liw) [inline src/pool/srv_target.c:1002]: [Minor] Here I think ds_pool_get is better.

**Review by liw — APPROVED** [review]

**Review by kccain — DISMISSED** [review]

**Comment by NiuYawei** [comment]
Addressed review comments, merged master to pick up CI trigger fix.

File: src/pool/srv_target.c [hunk 6]
```diff
@@ -1003,10 +1003,10 @@ ds_pool_tgt_connect(struct ds_pool *pool, struct pool_iv_conn *pic)
 	if (hdl == NULL)
 		D_GOTO(out, rc = -DER_NOMEM);
 
-	ds_pool_get(pool);
 	uuid_copy(hdl->sph_uuid, pic->pic_hdl);
 	hdl->sph_flags = pic->pic_flags;
 	hdl->sph_sec_capas = pic->pic_capas;
+	ds_pool_get(pool);
```
Comment (liw) [inline src/pool/srv_target.c:1009]: Thanks. Looks lovely.

**Review by liw — APPROVED** [review]

**Review by kccain — APPROVED** [review]

**Review by wangdi1 — APPROVED** [review]

**Review by Nasf-Fan — APPROVED** [review]

**Review by liw — APPROVED** [review]

File: src/pool/srv_target.c [hunk 7]
```diff
@@ -570,6 +570,21 @@ ds_pool_tgt_ec_eph_query_abort(struct ds_pool *pool)
...
+pool_fetch_hdls_ult_abort(struct ds_pool *pool)
+{
+	if (!pool->sp_fetch_hdls)
+		return;
+
+	ABT_mutex_lock(pool->sp_mutex);
+	ABT_cond_signal(pool->sp_fetch_hdls_cond);
+	ABT_mutex_unlock(pool->sp_mutex);
+
+	ABT_mutex_lock(pool->sp_mutex);
```
Comment (Nasf-Fan) [inline src/pool/srv_target.c:583]: It is not your patch issue, but I do not understand the purpose of unlock and lock again, it does not guarantee that others can obtain the mutex during such very short interval.

**Review by Nasf-Fan — APPROVED** [review]

File: src/pool/srv_target.c (hunk 7 above)
Comment (NiuYawei) [reply src/pool/srv_target.c:583]: That's required by ABT API, I suppose it's for avoiding race of "adding to wait queue" and "wakeup".

Anyway, I hope we could replace all these ad-hoc ULT abort functions with sched_req_wait() call.

**Pull request merged**